cmd/util/ledger/migrations/account_based_migration.go

```diff
@@ -0,0 +1,210 @@
+package migrations
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/pathfinder"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/util"
+)
+
+// PayloadToAccount takes a payload and return:
+// - (address, true, nil) if the payload is for an account, the account address is returned
+// - ("", false, nil) if the payload is not for an account
+// - ("", false, err) if running into any exception
+func PayloadToAccount(p ledger.Payload) (string, bool, error) {
+	k, err := p.Key()
+	if err != nil {
+		return "", false, fmt.Errorf("could not find key for payload: %w", err)
+	}
+	id, err := KeyToRegisterID(k)
+	if err != nil {
+		return "", false, fmt.Errorf("error converting key to register ID")
+	}
+	if len([]byte(id.Owner)) != flow.AddressLength {
+		return "", false, nil
+	}
+	return id.Owner, true, nil
+}
+
+// PayloadGroup groups payloads by account.
+// For global payloads, it's stored under NonAccountPayloads field
+type PayloadGroup struct {
+	NonAccountPayloads []ledger.Payload
+	Accounts           map[string][]ledger.Payload
+}
+
+// PayloadGrouping is a reducer function that adds the given payload to the corresponding
+// group under its account
+func PayloadGrouping(groups *PayloadGroup, payload ledger.Payload) (*PayloadGroup, error) {
+	address, isAccount, err := PayloadToAccount(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if isAccount {
+		groups.Accounts[address] = append(groups.Accounts[address], payload)
+	} else {
+		groups.NonAccountPayloads = append(groups.NonAccountPayloads, payload)
+	}
+
+	return groups, nil
+}
+
+// AccountMigrator takes all the payloads that belong to the given account
+// and return the migrated payloads
+type AccountMigrator func(account string, payloads []ledger.Payload, pathFinderVersion uint8) ([]ledger.Payload, error)
+
+// MigrateByAccount teaks a migrator function and all the payloads, and return the migrated payloads
+func MigrateByAccount(migrator AccountMigrator, allPayloads []ledger.Payload, pathFinderVersion uint8) (
```

Comment:
This function takes a migrator function, so that the caller can provide a different migrator function for different migration scenarios and reuse this function for running migration concurrently.